src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,123 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] dataElement;
+    private int size = 0;
+
+    public ArrayList() {
+        dataElement = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == dataElement.length) {
+            grow();
+        }
+        dataElement[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+
+        if (size == dataElement.length) {
+            grow();
+        }
 
+        System.arraycopy(dataElement, index, dataElement, index + 1, size - index);
+        dataElement[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null || list.isEmpty()) {
+            return;
+        }
 
+        ensureCapacity(size + list.size());
+
+        for (int i = 0; i < list.size(); i++) {
+            dataElement[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return dataElement[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        dataElement[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedIndex = dataElement[index];
+
+        System.arraycopy(dataElement, index + 1, dataElement, index, size - index - 1);
+        dataElement[--size] = null;
+
+        return removedIndex;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, dataElement[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("No such element: " + element);
```

Comment:
The exception message here is not very informative. It would be better to include more information about the error, such as the size of the list and the index that was attempted to be accessed.